crates/edr_provider/src/data.rs

```diff
@@ -1767,25 +1767,41 @@ impl<LoggerErrorT: Debug> ProviderData<LoggerErrorT> {
             &Box<dyn SyncState<StateError>>,
         ) -> T,
     ) -> Result<T, ProviderError<LoggerErrorT>> {
-        let maybe_block_number = self
-            .block_number_by_block_spec(block_spec.unwrap_or(&BlockSpec::Tag(BlockTag::Latest)))?;
-        if let Some(block_number) = maybe_block_number {
-            // We cannot use `LruCache::try_get_or_insert`, because it needs &mut self, but
-            // we would need &self in the callback to construct the context.
-            if !self.block_context_cache.contains(&block_number) {
+        let block = if let Some(block_spec) = block_spec {
+            self.block_by_block_spec(block_spec)?
+        } else {
+            Some(self.blockchain.last_block()?)
+        };
+
+        if let Some(block) = block {
+            let block_header = block.header();
+            let block_number = block_header.number;
+
+            // Avoid caching of states that may contain state overrides
+            let can_cache = self.irregular_state.state_overrides().is_empty();
```

Comment:
The original caching strategy wasn't taking into consideration the possibility of having irregular state changes to a block. If you made such changes, it would still return previously cached versions of the block.

My approach is too conservative and results in performance degradation for certain tests. I didn't have time before end-of-day to make this criteria more flexible though.

[This function](https://github.com/NomicFoundation/hardhat/blob/edr/main/crates/edr_evm/src/blockchain.rs#L208) shows how irregular state overrides affect the construction of a block.

Instead of just caching per block number, we can also incorporate the irregular state overrides that would impact the requested block number (everything up until this block number). I'm not sure how expensive it would be to validate this; e.g. using a hash; but you could experiment with this.

Alternatively, we could never cache irregular state. Even in that case we can make this condition less strict by looking whether there are an state overrides up until the requested block number.

Reply:
Thanks for the detailed explanation! I think something else might be going on here as well regarding the performance degradation. Running the Rocketpool scenario on my machine is almost 10x slower with this change than in the base branch, and Rocketpool doesn't use any `hardhat_*` methods which as I understand is the only way to set irregular state.

![image](https://github.com/NomicFoundation/hardhat/assets/5764438/122e3c84-ea1d-47ec-b4d5-39c5e7b0045e)

I created a flame graph using the Rockepool scenario on which two things stand out:

- Most of the time is spent in `compute_state_at_block`
- Dropping a snapshot takes significant time

I cherry picked this commit in the `edr/save-scenarios` branch so you can run the Rocketpool scenario yourself (after downloading it from the shared Google Drive) from the `crates/tools` directory:

`cargo run --release scenario ../../benchmark-scenarios/rocketpool-6a9dbfd8/06549782-8dc5-4700-bdde-9457cdd7b388`

And to generate a flame graph:

`cargo install flamegraph`

`CARGO_PROFILE_RELEASE_DEBUG=true cargo flamegraph --release -- scenario ../../benchmark-scenarios/rocketpool-6a9dbfd8/06549782-8dc5-4700-bdde-9457cdd7b388`

Make sure to ignore the part of the flame graph that loads the scenario files. This is easily identifiable as it uses rayon, not tokio.

Reply:
> I think something else might be going on here as well regarding the performance degradation. Running the Rocketpool scenario on my machine is almost 10x slower with this change than in the base branch, and Rocketpool doesn't use any hardhat_* methods which as I understand is the only way to set irregular state.

I'd need to look at the specific scenarios, but when forking a remote network with a configuration that has "genesis accounts", you're always creating an irregular state override for those, which is applied to that first block.

Additionally, we're always at least calculating the block state's once, even when using latest. This can be avoided by caching when we change the state.

I'm having a look at the scenario you highlighted to see whether either of these is the case and make the caching more aggressive.

Reply:
I made two optimisations:

1. I ensure that we always cache changes to a block, such that we can always use the latest version of the requested block number. This is the same behaviour as generating a state using `state_at_block_number.
2. I use a custom global allocator that retains memory pages such that we don't always have to go through the system allocator.

![image](https://github.com/NomicFoundation/hardhat/assets/6917585/aa21cf3b-5c58-45c2-81cc-b9d201cbe984)